features/FEATURE_LWIP/lwip-interface/LWIPMemoryManager.h

```diff
@@ -0,0 +1,141 @@
+/* Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef LWIP_MEMORY_MANAGER_H
+#define LWIP_MEMORY_MANAGER_H
+
+#include "EMACMemoryManager.h"
+
+class LWIPMemoryManager : public EMACMemoryManager {
+public:
+
+    /**
+     * Allocates memory buffer from the heap
+     *
+     * Memory buffer allocated from heap is always contiguous and can be arbitrary size.
+     *
+     * @param size     Size of the memory to allocate in bytes
+     * @param align    Memory alignment requirement in bytes
+     * @return         Allocated memory buffer, or NULL in case of error
+     */
+    virtual emac_mem_buf_t *alloc_heap(uint32_t size, uint32_t align);
+
+    /**
+     * Allocates memory buffer from a pool
+     *
+     * Memory buffer allocated from pool is contiguous if size is equal or less than
+     * (aligned) allocation unit, otherwise may be chained. Will typically come from
+     * fixed-size packet pool memory.
+     *
+     * @param  size    Size of the memory to allocate in bytes
+     * @param  align   Memory alignment requirement in bytes
+     * @return         Allocated memory buffer, or NULL in case of error
+     */
+    virtual emac_mem_buf_t *alloc_pool(uint32_t size, uint32_t align);
+
+    /**
+     * Get memory buffer pool allocation unit
+     *
+     * Returns the maximum size of contiguous memory that can be allocated from a pool.
+     *
+     * @param align    Memory alignment requirement in bytes
+     * @return         Contiguous memory size
+     */
+    virtual uint32_t get_pool_alloc_unit(uint32_t align);
+
+    /**
+     * Free memory buffer
+     *
+     * If memory buffer is chained must point to the start of the chain. Frees all buffers
+     * from the chained list.
+     *
+     * @param mem      Memory buffer to be freed.
+     */
+    virtual void free(emac_mem_buf_t *buf);
+
+    /**
+     * Return total length of a memory buffer
+     *
+     * Returns a length of this buffer and if the buffer is chained, the total length of
+     * following buffers in the list.
+     *
+     * @param mem      Memory buffer
+     * @return         Total length in bytes
+     */
+    virtual uint32_t get_total_len(emac_mem_buf_t *buf);
```

Comment:
Let's have some const-correctness - this should be a void buf pointer, and the method should be const.

Should all the methods be const apart from free and alloc -  if they only touch the buffer, they can't be changing the memory manager's overall state right? I suppose anything that modifies a buffer could be changing housekeeping info. So maybe const methods if they take const buf pointers.